### PR TITLE
Removed all pre-C++11 code, workarounds, and branches

### PR DESCRIPTION
**Pull request by seanm**
This removes a bunch of pre-C++11 workrarounds that are no longer necessary as C++11 is now required.

It is a nice clean up and simplification.

```
force_builders=Custom
build_image:Custom=centos:7
buildworker:Custom=linux-1
```

### CODE REVIEW

**Comment by vpisarev** [comment]
@seanm, I welcome those changes and believe something like this should be done. But maybe we should target 5.x first and do it slightly more aggressive, including C++ 14 and/or C++ 17 features.

Also, one of the things where some of our hacks should be removed in favor of a standard C++ feature is singletons. We have quite a few of them and could possibly replace them with std::call_once + std::once_flag.

@opencv-alalek, what do you think?

**Comment by seanm** [comment]
>But maybe we should target 5.x first...

Why?  4.x already has `#error "OpenCV 4.x+ requires enabled C++11 support"`.  This PR doesn't increase any requirements. All the code removed is already dead because of that existing `#error`.

>...do it slightly more aggressive, including C++ 14 and/or C++ 17 features

That's a whole other discussion. I'm not looking to drop C++11, I'm looking to drop C++03 and older.  Increasing the minimum language dialect requires wider discussion I'd imagine, but I'm not at all involved in this project's governance...

**Comment by vpisarev** [comment]
the major reason is that until 5.0 is released, we backport some patches to 3.4, which does not require C++ 11.

For 5.0 we already decided to make C++ 14 the minimum required standard. Maybe we will raise it up to C++ 17, we will see.

**Comment by seanm** [comment]
Well, I disagree, but it's up to you. I assume github allows you to merge this into either branch?

**Comment by seanm** [comment]
Do I have to rebase my commits onto the 5.x branch?  Or can everything be done from github?

**Comment by mshabunin** [comment]
This patch breaks windows builds and documentation build:
```
/build/precommit_docs/4.x/opencv/doc/tutorials/core/how_to_use_OpenCV_parallel_for_/how_to_use_OpenCV_parallel_for_.markdown:174: warning: block marked with [mandelbrot-parallel-call] for \snippet should appear twice in file how_to_use_OpenCV_parallel_for_.cpp, found it 0 times
/build/precommit_docs/4.x/opencv/doc/tutorials/core/how_to_use_OpenCV_parallel_for_new/how_to_use_OpenCV_parallel_for_new.markdown:107: warning: block marked with [convolution-parallel] for \snippet should appear twice in file how_to_use_OpenCV_parallel_for_new.cpp, found it 0 times
/build/precommit_docs/4.x/opencv/doc/tutorials/core/how_to_use_OpenCV_parallel_for_new/how_to_use_OpenCV_parallel_for_new.markdown:114: warning: block marked with [overload-full] for \snippet should appear twice in file how_to_use_OpenCV_parallel_for_new.cpp, found it 0 times
/build/precommit_docs/4.x/opencv/doc/tutorials/core/how_to_use_OpenCV_parallel_for_new/how_to_use_OpenCV_parallel_for_new.markdown:117: warning: block marked with [convolution-parallel-function] for \snippet should appear twice in file how_to_use_OpenCV_parallel_for_new.cpp, found it 0 times
/build/precommit_docs/4.x/opencv/doc/tutorials/core/how_to_use_OpenCV_parallel_for_new/how_to_use_OpenCV_parallel_for_new.markdown:122: warning: block marked with [overload-row-split] for \snippet should appear twice in file how_to_use_OpenCV_parallel_for_new.cpp, found it 0 times
/build/precommit_docs/4.x/opencv/doc/tutorials/core/how_to_use_OpenCV_parallel_for_new/how_to_use_OpenCV_parallel_for_new.markdown:125: warning: block marked with [convolution-parallel-function-row] for \snippet should appear twice in file how_to_use_OpenCV_parallel_for_new.cpp, found it 0 times
```

**Comment by seanm** [comment]
> This patch breaks windows builds and documentation build:

I think I've just fixed the doc build...

**Comment by mshabunin** [comment]
There are still issues with Windows build (Visual Studio 2019 Developer Command Prompt v16.11.21)
```
C:\build\precommit_windows64\4.x\opencv\modules\core\include\opencv2/core/cvdef.h(752): 
fatal error C1189: #error:  "OpenCV 4.x+ requires enabled C++11 support" 
(compiling source file C:\build\precommit_windows64\4.x\opencv\modules\imgproc\src\imgwarp.avx2.cpp) 
```
Perhaps we need some special handling for MSVC - https://devblogs.microsoft.com/cppblog/msvc-now-correctly-reports-__cplusplus/ , but if we choose this way, we can still break user's aplications with our `cvdef.hpp`. So alternatively we need to use old `_MSVC_LANG` macro (https://learn.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=msvc-170#microsoft-specific-predefined-macros).

**Comment by seanm** [comment]
> Perhaps we need some special handling for MSVC - https://devblogs.microsoft.com/cppblog/msvc-now-correctly-reports-__cplusplus/ , but if we choose this way, we can still break user's aplications with our `cvdef.hpp`. So alternatively we need to use old `_MSVC_LANG` macro (https://learn.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=msvc-170#microsoft-specific-predefined-macros). Example from vulkan header:

Interesting.  I've updated my commit with something like the vulkan header.

I don't have Windows though, so this may take me a few trys with the help of github CI...

File: modules/core/include/opencv2/core/cvdef.h [hunk 1]
```diff
@@ -744,89 +744,42 @@ __CV_ENUM_FLAGS_BITWISE_XOR_EQ   (EnumType, EnumType)
...
-#    define CV_CXX11 1
-#  endif
-#else
-#  if CV_CXX11 == 0
-#    undef CV_CXX11
+#ifdef __cplusplus
+// MSVC was stuck at __cplusplus == 199711L for a long time, even where it supports C++11,
+// so check _MSVC_LANG instead. See:
+// <https://devblogs.microsoft.com/cppblog/msvc-now-correctly-reports-__cplusplus>
+#  if defined(_MSVC_LANG) && (_MSVC_LANG < 201103L)
```
Comment (mshabunin) [inline modules/core/include/opencv2/core/cvdef.h:754]: Something is still wrong with VS 2015. Maybe change it to MSVC version check (removed line 773): `defined(_MSC_VER) && _MSC_VER < 1800`

Comment (seanm) [reply modules/core/include/opencv2/core/cvdef.h:754]: Done.

File: modules/core/include/opencv2/core/cvdef.h [hunk 2]
```diff
@@ -744,89 +744,42 @@ __CV_ENUM_FLAGS_BITWISE_XOR_EQ   (EnumType, EnumType)
...
-#else
-#  if CV_CXX11 == 0
-#    undef CV_CXX11
+#ifdef __cplusplus
+// MSVC was stuck at __cplusplus == 199711L for a long time, even where it supports C++11,
+// so check _MSVC_LANG instead. See:
+// <https://devblogs.microsoft.com/cppblog/msvc-now-correctly-reports-__cplusplus>
+#  if defined(_MSC_VER) && (_MSC_VER < 1800)
+#    error "OpenCV 4.x+ requires enabled C++11 support"
+#  elif __cplusplus < 201103L
```
Comment (mshabunin) [inline modules/core/include/opencv2/core/cvdef.h:756]: It still does not work, first condition passes, but second condition still fails.
```
C:\build\precommit_windows64\4.x\opencv\modules\core\include\opencv2/core/cvdef.h(757): fatal error C1189: #error:  "OpenCV 4.x+ requires enabled C++11 support" 
```
Should be something like this:
```.cpp
#if GOOD_MSVC_VERSION
 // ok
#elif BAD_MSVC_VERSION || BAD_CPLUSPLUS_VERSION
 // error
#endif
```
How can it be optimized :thinking: ?

Comment (mshabunin) [reply modules/core/include/opencv2/core/cvdef.h:756]: Maybe:
```.cpp
#if MSVC
  #if BAD_VERSION
    // error
  #endif
#elif BAD_CPLUSPLUS
  // error
#endif
```

Comment (seanm) [reply modules/core/include/opencv2/core/cvdef.h:756]: Yes, that makes sense; done. Sorry this is taking a few iterations, like I said, I don't have MSVC locally. :)

**Comment by mshabunin** [comment]
Good, `ifdef` part is passing now, but several compilation issues have risen in `cap_msmf` and some _opencv_contrib_ modules.
https://pullrequest.opencv.org/buildbot/builders/precommit_windows64/builds/102452/steps/compile%20release/logs/stdio
:face_exhaling: I hope that most of them can be resolved by including correct headers or minor modifications, but there is possibility that some parts need to be rewritten to comply with this picky compiler and its C++ standard vision :slightly_smiling_face: .

**Comment by seanm** [comment]
> Good, `ifdef` part is passing now, but several compilation issues have risen in `cap_msmf` and some _opencv_contrib_ modules. https://pullrequest.opencv.org/buildbot/builders/precommit_windows64/builds/102452/steps/compile%20release/logs/stdio 😮‍💨 I hope that most of them can be resolved by including correct headers or minor modifications, but there is possibility that some parts need to be rewritten to comply with this picky compiler and its C++ standard vision 🙂 .

Seems most are due to missing headers, due to me removing the <array> include in cvdef.h.  I'll fix them.

But some are from `opencv_contrib` which seems to be a different repro. I can fix them too, but I wonder how much other code out there is expecting <array> to be automatically included...

**Comment by seanm** [comment]
Actually, by including `<array>` in the appropriate .pch files I think I've fixed the remaining issues... CI seems to be green now.

**Review by vpisarev — APPROVED** [review]

File: doc/tutorials/core/how_to_use_OpenCV_parallel_for_new/how_to_use_OpenCV_parallel_for_new.markdown [hunk 3]
```diff
@@ -96,36 +96,10 @@ When looking at the sequential implementation, we can notice that each pixel dep
...
-2. Split the rows and compute for each row:
-
-    @snippet how_to_use_OpenCV_parallel_for_new.cpp overload-row-split
-
-    In this case, we call the parallel_for_ function with a different range:
-    @snippet how_to_use_OpenCV_parallel_for_new.cpp convolution-parallel-function-row
-
-@note In our case, both implementations perform similarly. Some cases may allow better memory access patterns or other performance benefits.
-
-To set the number of threads, you can use: @ref cv::setNumThreads. You can also specify the number of splitting using the nstripes parameter in @ref cv::parallel_for_. For instance, if your processor has 4 threads, setting `cv::setNumThreads(2)` or setting `nstripes=2` should be the same as by default it will use all the processor threads available but will split the workload only on two threads.
```
[the remark this answers is not in the document]
Comment (seanm) [reply doc/tutorials/core/how_to_use_OpenCV_parallel_for_new/how_to_use_OpenCV_parallel_for_new.markdown:123]: Done.

File: doc/tutorials/core/how_to_use_OpenCV_parallel_for_/how_to_use_OpenCV_parallel_for_.markdown [hunk 4]
```diff
@@ -165,21 +165,10 @@ This splitting is done automatically to distribute equally the computation load.
...
-The parallel execution is called with:
+The parallel execution is called with a lambda expression:
 
 @snippet how_to_use_OpenCV_parallel_for_.cpp mandelbrot-parallel-call
 
-Here, the range represents the total number of operations to be executed, so the total number of pixels in the image.
-To set the number of threads, you can use: @ref cv::setNumThreads. You can also specify the number of splitting using the
-nstripes parameter in @ref cv::parallel_for_. For instance, if your processor has 4 threads, setting `cv::setNumThreads(2)`
-or setting `nstripes=2` should be the same as by default it will use all the processor threads available but will split the
-workload only on two threads.
```
[the remark this answers is not in the document]
Comment (seanm) [reply doc/tutorials/core/how_to_use_OpenCV_parallel_for_/how_to_use_OpenCV_parallel_for_.markdown:176]: Done.

**Comment by asmorkalov** [comment]
Documentation warnings on Buildbot:
```
/build/precommit_docs/4.x/opencv/doc/tutorials/core/how_to_use_OpenCV_parallel_for_new/how_to_use_OpenCV_parallel_for_new.markdown:114: warning: block marked with [overload-full] for \snippet should appear twice in file how_to_use_OpenCV_parallel_for_new.cpp, found it 0 times
/build/precommit_docs/4.x/opencv/doc/tutorials/core/how_to_use_OpenCV_parallel_for_new/how_to_use_OpenCV_parallel_for_new.markdown:117: warning: block marked with [convolution-parallel-function] for \snippet should appear twice in file how_to_use_OpenCV_parallel_for_new.cpp, found it 0 times
/build/precommit_docs/4.x/opencv/doc/tutorials/core/how_to_use_OpenCV_parallel_for_new/how_to_use_OpenCV_parallel_for_new.markdown:122: warning: block marked with [overload-row-split] for \snippet should appear twice in file how_to_use_OpenCV_parallel_for_new.cpp, found it 0 times
/build/precommit_docs/4.x/opencv/doc/tutorials/core/how_to_use_OpenCV_parallel_for_new/how_to_use_OpenCV_parallel_for_new.markdown:125: warning: block marked with [convolution-parallel-function-row] for \snippet should appear twice in file how_to_use_OpenCV_parallel_for_new.cpp, found it 0 times
```

**Comment by seanm** [comment]
> Documentation warnings on Buildbot:

Yeah, that's because I brought back the comments you asked me to. :)  CI was green before.

File: doc/tutorials/core/how_to_use_OpenCV_parallel_for_new/how_to_use_OpenCV_parallel_for_new.markdown [hunk 5]
```diff
@@ -96,8 +96,8 @@ When looking at the sequential implementation, we can notice that each pixel dep
 
 @note Although values of a pixel in a particular stripe may depend on pixel values outside the stripe, these are only read only operations and hence will not cause undefined behaviour.
 
+C++11 standard allows a parallel implementation with a lambda expression:
```
Comment (asmorkalov) [inline doc/tutorials/core/how_to_use_OpenCV_parallel_for_new/how_to_use_OpenCV_parallel_for_new.markdown:99]: The following documentation describes `operator()` overloading and other steps of the old version. I propose not to remove pre-C++11 code, but just remove ifdefs for C++11.

File: samples/cpp/tutorial_code/core/how_to_use_OpenCV_parallel_for_/how_to_use_OpenCV_parallel_for_new.cpp [hunk 6]
```diff
@@ -116,126 +115,8 @@ void conv_parallel_row_split(Mat src, Mat &dst, Mat kernel)
...
-    int sz;
-
-public:
-    parallelConvolution(Mat src, Mat &dst, Mat kernel)
-        : m_src(src), m_dst(dst), m_kernel(kernel)
-    {
-        sz = kernel.rows / 2;
-    }
-
-    //! [overload-full]
```
Comment (asmorkalov) [inline samples/cpp/tutorial_code/core/how_to_use_OpenCV_parallel_for_/how_to_use_OpenCV_parallel_for_new.cpp:138]: There is reference to `overload-full` block in markdown.

File: samples/cpp/tutorial_code/core/how_to_use_OpenCV_parallel_for_/how_to_use_OpenCV_parallel_for_new.cpp [hunk 7]
```diff
@@ -116,126 +115,8 @@ void conv_parallel_row_split(Mat src, Mat &dst, Mat kernel)
...
-    int sz = kernel.rows / 2;
-    copyMakeBorder(src, src, sz, sz, sz, sz, BORDER_REPLICATE);
-
-    //! [convolution-parallel-function]
-    parallelConvolution obj(src, dst, kernel);
-    parallel_for_(Range(0, rows * cols), obj);
-    //! [convolution-parallel-function]
-}
-
-//! [conv-parallel-row-split]
```
Comment (asmorkalov) [inline samples/cpp/tutorial_code/core/how_to_use_OpenCV_parallel_for_/how_to_use_OpenCV_parallel_for_new.cpp:179]: The same.

File: samples/cpp/tutorial_code/core/how_to_use_OpenCV_parallel_for_/how_to_use_OpenCV_parallel_for_new.cpp [hunk 8]
```diff
@@ -116,126 +115,8 @@ void conv_parallel_row_split(Mat src, Mat &dst, Mat kernel)
...
-
-    dst = Mat(rows, cols, CV_8UC1, Scalar(0));
-
-    // Taking care of edge values
-    // Make border = kernel.rows / 2;
-
-    int sz = kernel.rows / 2;
-    copyMakeBorder(src, src, sz, sz, sz, sz, BORDER_REPLICATE);
-
-    //! [convolution-parallel-function]
```
Comment (asmorkalov) [inline samples/cpp/tutorial_code/core/how_to_use_OpenCV_parallel_for_/how_to_use_OpenCV_parallel_for_new.cpp:173]: The same,

File: samples/cpp/tutorial_code/core/how_to_use_OpenCV_parallel_for_/how_to_use_OpenCV_parallel_for_new.cpp [hunk 9]
```diff
@@ -116,126 +115,8 @@ void conv_parallel_row_split(Mat src, Mat &dst, Mat kernel)
...
-    int sz;
-
-public:
-    parallelConvolutionRowSplit(Mat src, Mat &dst, Mat kernel)
-        : m_src(src), m_dst(dst), m_kernel(kernel)
-    {
-        sz = kernel.rows / 2;
-    }
-
-    //! [overload-row-split]
```
Comment (asmorkalov) [inline samples/cpp/tutorial_code/core/how_to_use_OpenCV_parallel_for_/how_to_use_OpenCV_parallel_for_new.cpp:194]: The same. See CI warnings.

**Comment by asmorkalov** [comment]
@seanm friendly reminder.

**Comment by seanm** [comment]
> @seanm friendly reminder.

I don't understand what it is you want to do in those documentation files. And I'm not familiar with the file's syntax nor the meaning of those tags. Nor am I familiar with the classes the document is describing. I think I give up at this point. Feel free to take what I've started and continue it.

File: doc/tutorials/core/how_to_use_OpenCV_parallel_for_/how_to_use_OpenCV_parallel_for_.markdown [hunk 10]
```diff
@@ -3,12 +3,13 @@ How to use the OpenCV parallel_for_ to parallelize your code {#tutorial_how_to_u
 
 @tableofcontents
 
-@prev_tutorial{tutorial_file_input_output_with_xml_yml}
-
```
Comment (asmorkalov) [inline doc/tutorials/core/how_to_use_OpenCV_parallel_for_/how_to_use_OpenCV_parallel_for_.markdown]: To be resored.

File: doc/tutorials/core/how_to_use_OpenCV_parallel_for_/how_to_use_OpenCV_parallel_for_.markdown [hunk 11]
```diff
@@ -3,12 +3,13 @@ How to use the OpenCV parallel_for_ to parallelize your code {#tutorial_how_to_u
...
 |    |    |
 | -: | :- |
 | Compatibility | OpenCV >= 3.0 |
 
+@warning
+This tutorial can contain obsolete information. Please check modern version here: [tutorial_how_to_use_OpenCV_parallel_for_new]
```
Comment (asmorkalov) [inline doc/tutorials/core/how_to_use_OpenCV_parallel_for_/how_to_use_OpenCV_parallel_for_.markdown]: link is broken

**Review by asmorkalov — APPROVED** [review]

**Comment by mshabunin** [comment]
@asmorkalov , `tutorial_how_to_use_OpenCV_parallel_for_` is considered obsolete and is kept for only for link preservation (as I understand). It is not listed in the corresponding section: https://docs.opencv.org/4.x/de/d7a/tutorial_table_of_content_core.html

`tutorial_how_to_use_OpenCV_parallel_for_new` is the new and updated tutorial.

See #20361

**Comment by seanm** [comment]
I thought you didn't want this merged until 5.x?

**Comment by mshabunin** [comment]
3.4 branch is considered EOL now (https://github.com/opencv/opencv/wiki/Branches), so we don't need to do backporting anymore (it was the main concern originally). And forward porting to 5.x will be done by the OpenCV team on a regular basis.

I believe there is still no decision on minimal C++ standard for 5.x.

**Pull request merged**